### PR TITLE
Feat/golang/v2 bindings

### PR DESCRIPTION
**Pull request by nonam3e**
## Describe the changes

This PR updates current golang bindings to icicle v2

### CODE REVIEW

File: wrappers/golang/curves/bls12377/ecntt.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+//go:build ecntt
```
Comment (jeremyfelder) [inline wrappers/golang/curves/bls12377/ecntt.go:1]: I think its best to remove the build tags for g2 and ecntt and move these functionalities to separate packages underneath each curve package…apparently its an "abuse" of build tags to add API functionality and we didn’t know this when we first added G2; see [this comment](https://github.com/golang/go/issues/43288#issuecomment-749152792) and discussion

**Review by jeremyfelder — COMMENTED** [review]
There are changes needed in the [build.sh](https://github.com/ingonyama-zk/icicle/pull/400/files#diff-99cb40b29676fcd123df8c227a952343014510910a7812736b99557e14d7fcb9) script and the [CI](https://github.com/ingonyama-zk/icicle/pull/400/files#diff-e7f2e14bd868dd64c141ea94986567083b2100d35faac4e1ad66c772a41228c7) to work with V2 structure.

We should let the checks run to ensure that the changes don't break builds and tests.

There will need to be a few more changes to fully support the new v2 structure once golang supports hash or field only packages but I guess this can be done once we introduce a field or hash binding

File: wrappers/golang/README.md [hunk 2]
```diff
@@ -5,19 +5,21 @@ In order to build the underlying ICICLE libraries you should run the build scrip
...
 
 ```
-./build <curve> [G2_enabled]
+./build <curve> [-cuda_version=<version>] [-g2] [-ecntt] [-devmode]
 
 curve - The name of the curve to build or "all" to build all curves
-G2_enabled - Optional - To build with G2 enabled 
+-g2 - Optional - To build with G2 enabled 
+-ecntt - Optional - To build with ECNTT enabled
+-devmode - Optional - To build in devmode
```
Comment (jeremyfelder) [inline wrappers/golang/README.md:13]:
```suggestion
-g2 - Optional - Build with G2 enabled 
-ecntt - Optional - Build with ECNTT enabled
-devmode - Optional - Build in devmode
```

File: wrappers/golang/build.sh [hunk 3]
```diff
@@ -2,32 +2,59 @@
...
+            ;;
+        -ecntt*)
+            # ECNTT_DEFINED=$(echo "$arg" | cut -d'=' -f2)
+            ECNTT_DEFINED=ON
+            ;;
+        -g2*)
+            # G2_DEFINED=$(echo "$arg" | cut -d'=' -f2)
+            G2_DEFINED=ON
+            ;;
+        -devmode*)
```
Comment (jeremyfelder) [inline wrappers/golang/build.sh:36]: We want specific flags and not to match a flag with a prefix

```suggestion
        -ecntt)
            # ECNTT_DEFINED=$(echo "$arg" | cut -d'=' -f2)
            ECNTT_DEFINED=ON
            ;;
        -g2)
            # G2_DEFINED=$(echo "$arg" | cut -d'=' -f2)
            G2_DEFINED=ON
            ;;
        -devmode)
```

File: wrappers/golang/curves/bls12377/ecntt/ecntt_test.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	. "github.com/ingonyama-zk/icicle/wrappers/golang/curves/bls12377"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	largestTestSize = 17
+)
+
+func TestECNtt(t *testing.T) {
```
Comment (jeremyfelder) [inline wrappers/golang/curves/bls12377/ecntt/ecntt_test.go:16]: Since this is a separate package from bls12377, this test needs its own `initDomain` call since the bls12377 package will releaseDomain in the teardown portion of its tests.

File: wrappers/golang/curves/bls12377/ecntt/include/ecntt.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#include <cuda_runtime.h>
+#include "../../../include/types.h"
+#include <stdbool.h>
```
Comment (jeremyfelder) [inline wrappers/golang/curves/bls12377/ecntt/include/ecntt.h:3]: This is not needed

File: wrappers/golang/curves/bls12377/g2/g2_msm.go [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
-
-package bls12377
+package g2
 
 // #cgo CFLAGS: -I./include/
 // #include "g2_msm.h"
 import "C"
 
 import (
+	. "github.com/ingonyama-zk/icicle/wrappers/golang/curves/bls12377"
```
Comment (jeremyfelder) [inline wrappers/golang/curves/bls12377/g2/g2_msm.go:8]: Why is this needed?

File: wrappers/golang/curves/bls12377/g2/g2_msm_test.go [hunk 7]
```diff
@@ -11,6 +9,7 @@ import (
 
 	"github.com/ingonyama-zk/icicle/wrappers/golang/core"
 	cr "github.com/ingonyama-zk/icicle/wrappers/golang/cuda_runtime"
+	. "github.com/ingonyama-zk/icicle/wrappers/golang/curves/bls12377"
```
Comment (jeremyfelder) [inline wrappers/golang/curves/bls12377/g2/g2_msm_test.go:12]: Why is this needed?

File: wrappers/golang/curves/bls12377/g2/g2_msm.go (hunk 6 above)
Comment (nonam3e) [reply wrappers/golang/curves/bls12377/g2/g2_msm.go:8]: it's needed to import `ScalarField` and its functions

File: wrappers/golang/internal/generator/templates/ecntt_test.go.tmpl [hunk 8]
```diff
@@ -0,0 +1,49 @@
+package ecntt
+
+import (
+	"testing"
+
+	"github.com/consensys/gnark-crypto/ecc/bn254/fr/fft"
```
Comment (jeremyfelder) [inline wrappers/golang/internal/generator/templates/ecntt_test.go.tmpl:6]: This should be curve specific
```suggestion
	"github.com/consensys/gnark-crypto/ecc/{{.GnarkImport}}/fr/fft"
```
